include/srf/experimental/modules/segment_modules.hpp

```diff
@@ -0,0 +1,173 @@
+    using segment_module_typeinfo_map_t = std::map<std::string, const std::type_info*>;
+
+    SegmentModule() = delete;
+    SegmentModule(std::string module_name);
+    SegmentModule(std::string module_name, nlohmann::json config);
+
+    const std::string& component_prefix() const;
+    const nlohmann::json& config() const;
+    const std::string& name() const;
+
+    std::string get_module_component_name(const std::string& component_name) const;
+
+    /**
+     * Return vector of input names -- these are only understood by the SegmentModule
+     * @return std::vector
+     */
+    const std::vector<std::string> input_ids() const;
+
+    /**
+     * Return a vector of output names -- these are only understood by the SegmentModule class
+     * @return std::vector
+     */
+    const std::vector<std::string> output_ids() const;
+
+    /**
+     * Return a set of ObjectProperties for module input_ids
+     * @return ObjectProperties
+     */
+    const segment_module_port_map_t input_ports() const;
+
+    /**
+     * Return the ObjectProperties object corresponding to input_name
+     * @param input_name Name of the module port
+     * @return ObjectProperties
+     */
+    segment_module_port_t input_port(const std::string& input_name);
+
+    /**
+     * Return a map of module port id : type indices
+     * @return std::map
+     */
+    const std::map<std::string, const std::type_info*> input_port_type_ids() const;
+
+    /**
+     * Return the type index of a given input name
+     * @return Type index
+     */
+    const std::type_info* input_port_type_id(const std::string& input_name);
+
+    /**
+     * Return a set of ObjectProperties for module input_ids
+     * @return ObjectProperties
+     */
+    const segment_module_port_map_t output_ports() const;
+
+    /**
+     * Return an ObjectProperties for the module port corresponding to output name
+     * @param output_name Name of the module port to return
+     * @return ObjectProperties
+     */
+    segment_module_port_t output_port(const std::string& output_name);
+
+    /**
+     * Return a map of module port id : type indices
+     * @return std::map
+     */
+    const segment_module_typeinfo_map_t output_port_type_ids() const;
+
+    /**
+     * Return the type index of a given input name
+     * @param input_name Name of the module port to return a type index for
+     * @return Type index
+     */
+    const std::type_info* output_port_type_id(const std::string& output_name);
+
+    /**
+     * Functional entrypoint for module constructor during build -- this lets us act like a std::function
+     * @param builder
+     */
+    void operator()(segment::Builder& builder)
+    {
+        this->initialize(builder);
+    };
+
+    /* Virtual Functions */
+    /**
+     * Entrypoint for module constructor during build
+     * @param builder
+     */
+    virtual void initialize(segment::Builder& builder) = 0;
```

Comment:
So if I understand correctly this just defines the API of what a module should look like, and outside of the impls in the unittests we won't have any concrete subclasses of this in SRF?